src/Features/Core/Portable/EditAndContinue/EditAndContinueWorkspaceService.cs

```diff
@@ -387,18 +384,38 @@ public Task<bool> HasChangesAsync(Solution solution, SolutionActiveStatementSpan
                 editSession.StorePendingUpdate(solution, solutionUpdate);
             }
 
-            // clear emit/apply diagnostics reported previously:
-            _emitDiagnosticsUpdateSource.ClearDiagnostics();
+            // Note that we may return empty deltas if all updates have been deferred.
+            // The debugger will still call commit or discard on the update batch.
+            return (solutionUpdate.Summary, solutionUpdate.Deltas, ToDiagnosticData(solution, solutionUpdate.Diagnostics));
+        }
 
-            // report emit/apply diagnostics:
-            foreach (var (projectId, diagnostics) in solutionUpdate.Diagnostics)
+        private static ImmutableArray<DiagnosticData> ToDiagnosticData(Solution solution, ImmutableArray<(ProjectId ProjectId, ImmutableArray<Diagnostic> Diagnostics)> diagnosticsByProject)
+        {
+            var _ = ArrayBuilder<DiagnosticData>.GetInstance(out var result);
+
+            foreach (var (projectId, diagnostics) in diagnosticsByProject)
             {
-                _emitDiagnosticsUpdateSource.ReportDiagnostics(_workspace, solution, projectId, diagnostics);
+                var project = solution.GetProject(projectId);
+
+                foreach (var diagnostic in diagnostics)
+                {
+                    var document = solution.GetDocument(diagnostic.Location.SourceTree);
```

Comment:
tagging @jasonmalinowski more cases of this.  we have to consider SGs and waht is ok.

@tmat this doesn't affect you.  this is related to a  covno jason and i are having about this API and what guarantees there are around it.